File: code/R/gamma.R

# ガンマ分布 -------------------------------------------------------------------

# 利用するパッケージ
library(tidyverse)
library(gganimate)


# 確率の計算 -------------------------------------------------------------------

# パラメータを指定
a <- 2
b <- 2

# 確率変数の値を指定
lambda <- 2


# 定義式により確率密度を計算
C <- b^a / gamma(a)
dens <- C * lambda^(a - 1) * exp(-b * lambda)
dens

# 対数をとった定義式により確率密度を計算
log_C <- a * log(b) - lgamma(a)
log_dens <- log_C + (a - 1) * log(lambda) - b * lambda
dens <- exp(log_dens)
dens; log_dens

# ガンマ分布の関数により確率密度を計算
dens <- dgamma(x = lambda, shape = a, rate = b)
dens

# ガンマ分布の関数により確率密度を計算
dens <- dgamma(x = lambda, shape = a, scale = 1 / b)
dens

# ガンマ分布の対数をとった関数により確率密度を計算
log_dens <- dgamma(x = lambda, shape = a, rate = b, log = TRUE)
dens <- exp(log_dens)
dens; log_dens

# ガンマ分布の対数をとった関数により確率密度を計算
log_dens <- dgamma(x = lambda, shape = a, scale = 1 / b, log = TRUE)
dens <- exp(log_dens)
dens; log_dens


# 統計量の計算 ------------------------------------------------------------------

# パラメータを指定
a <- 2
b <- 2


# 平均を計算
E_lambda <- a / b
E_lambda

# 分散を計算
V_lambda <- a / b^2
V_lambda

# 最頻値を計算
mode_lambda <- (a - 1) / b
mode_lambda


# 分布の可視化 ------------------------------------------------------------------

# パラメータを指定
a <- 2
b <- 2

# 作図用のlambdaの点を作成
lambda_vals <- seq(from = 0, to = 5, length.out = 250)

# ガンマ分布を計算
dens_df <- tidyr::tibble(
  lambda = lambda_vals, # 確率変数
  density = dgamma(x = lambda_vals, shape = a, rate = b) # 確率密度
)


# ガウス分布を作図
ggplot(data = dens_df, mapping = aes(x = lambda, y = density)) + # データ
  geom_line(color = "#00A968") + # 折れ線グラフ
  labs(title = "Gamma Distribution", 
       subtitle = paste0("a=", a, ", b=", b), 
       x = expression(lambda)) # ラベル


# 統計量の計算
E_lambda <- a / b
s_lambda <- sqrt(a / b^2)
mode_lambda <- (a - 1) / b

# 統計量を重ねたガウス分布を作図
ggplot(data = dens_df, mapping = aes(x = lambda, y = density)) + # データ
  geom_line(color = "#00A968") + # 分布
  geom_vline(xintercept = E_lambda, color = "orange", size = 1, linetype = "dashed") + # 平均
  geom_vline(xintercept = E_lambda - s_lambda, color = "orange", size = 1, linetype = "dotted") + # 平均 - 標準偏差
  geom_vline(xintercept = E_lambda + s_lambda, color = "orange", size = 1, linetype = "dotted") + # 平均 + 標準偏差
  geom_vline(xintercept = mode_lambda, color = "chocolate", size = 1, linetype = "dashed") + # 最頻値
  labs(title = "Gamma Distribution", 
       subtitle = paste0("a=", a, ", b=", b), 
       x = expression(lambda)) # ラベル


# パラメータと分布の形状の関係 ----------------------------------------------------------

# パラメータとして利用する値を指定
a_vals <- seq(from = 0.1, to = 10, by = 0.1)
b_vals <- seq(from = 0.1, to = 10, by = 0.1)

# 固定するパラメータを指定
a <- 2
b <- 2

# 作図用のlambdaの点を作成
lambda_vals <- seq(from = 0, to = 5, length.out = 250)

# a, bの値ごとに分布を計算
anime_dens_df <- tidyr::tibble()
#for(a in a_vals) {
for(b in b_vals) {
  # ガンマ分布を計算
  tmp_dens_df <- tidyr::tibble(
    lambda = lambda_vals, # 確率変数
    density = dgamma(x = lambda_vals, shape = a, rate = b), # 確率密度
    parameter = paste0("a=", a, ", b=", b) %>% 
      as.factor() # フレーム切替用のラベル
  )
  
  # 結果を結合
  anime_dens_df <- rbind(anime_dens_df, tmp_dens_df)
}

# アニメーション用のガウス分布を作図
anime_dens_graph <- ggplot(data = anime_dens_df, mapping = aes(x = lambda, y = density)) + # データ
  geom_line(color = "#00A968") + # 折れ線グラフ
  gganimate::transition_manual(parameter) + # フレーム
  labs(title = "Gamma Distribution", 
       subtitle = "{current_frame}", 
       x = expression(lambda)) # ラベル

# gif画像を作成
gganimate::animate(anime_dens_graph, nframes = length(a_vals), fps = 100)


# 乱数の生成 -------------------------------------------------------------------

### 乱数を生成 -----

# パラメータを指定
a <- 2
b <- 2

# データ数(サンプルサイズ)を指定
N <- 1000

# ガンマ分布に従う乱数を生成
lambda_n <- rgamma(n = N, shape = a, rate = b)

# サンプルを格納
data_df <- tidyr::tibble(lambda = lambda_n)

# 作図用のlambdaの点を作成
lambda_vals <- seq(from = 0, to = max(lambda_n) + 1, length.out = 250)

# ガンマ分布を計算
dens_df <- tidyr::tibble(
  lambda = lambda_vals, # 確率変数
  density = dgamma(x = lambda_vals, shape = a, rate = b) # 確率密度
)


### 乱数の可視化 -----

# サンプルの頻度を作図
ggplot() + 
  geom_histogram(data = data_df, mapping = aes(x = lambda), 
                 breaks = seq(from = min(lambda_vals), to = max(lambda_vals), length.out = 30), 
                 fill = "#00A968", color = "#00A968") + # ヒストグラム
  labs(title = "Gamma Distribution", 
       subtitle = paste0("a=", a, ", b=", b), 
       x = expression(lambda)) # ラベル

# サンプルの密度を作図
ggplot() + 
  geom_histogram(data = data_df, mapping = aes(x = lambda, y = ..density..), 
                 breaks = seq(from = min(lambda_vals), to = max(lambda_vals), length.out = 30), 
                 fill = "#00A968", color = "#00A968") + # ヒストグラム
  geom_line(data = dens_df, mapping = aes(x = lambda, y = density), 
            color = "darkgreen", linetype = "dashed") + # 元の分布
  labs(title = "Gamma Distribution", 
       subtitle = paste0("a=", a, ", b=", b), 
       x = expression(lambda)) # ラベル


### アニメーションによる可視化 -----

# フレーム数を指定
N_frame <- 100

# 乱数を1つずつ生成
lambda_n <- rep(NA, times = N_frame)
anime_freq_df <- tidyr::tibble()
anime_data_df <- tidyr::tibble()
anime_dens_df <- tidyr::tibble()
for(n in 1:N_frame) {
  # ガンマ分布に従う乱数を生成
  lambda_n[n] <- rgamma(n = 1, shape = a, rate = b)
  
  # ラベル用のテキストを作成
  label_text <- paste0("a=", a, ", b=", b, ", N=", n)
  
  # n個の乱数を格納
  tmp_freq_df <- tidyr::tibble(
    lambda = lambda_n[1:n], # サンプル
    parameter = as.factor(label_text) # フレーム切替用のラベル
  )
  
  # n番目の乱数を格納
  tmp_data_df <- tidyr::tibble(
    lambda = lambda_n[n], # サンプル
    parameter = as.factor(label_text) # フレーム切替用のラベル
  )
  
  # n回目のラベルを付与
  tmp_dens_df <- dens_df %>% 
    dplyr::mutate(parameter = as.factor(label_text))
  
  # 結果を結合
  anime_freq_df <- rbind(anime_freq_df, tmp_freq_df)
  anime_data_df <- rbind(anime_data_df, tmp_data_df)
  anime_dens_df <- rbind(anime_dens_df, tmp_dens_df)
}


# アニメーション用のサンプルの頻度を作図
anime_freq_graph <- ggplot() + 
  geom_histogram(data = anime_freq_df, mapping = aes(x = lambda), 
                 breaks = seq(from = min(lambda_vals), to = max(lambda_vals), length.out = 30), 
                 fill = "#00A968", color = "#00A968") + # ヒストグラム
  geom_point(data = anime_data_df, mapping = aes(x = lambda, y = 0), 
             color = "orange", size = 5) + # サンプル
  gganimate::transition_manual(parameter) + # フレーム
  labs(title = "Gamma Distribution", 
       subtitle = "{current_frame}") # ラベル

# gif画像を作成
gganimate::animate(anime_freq_graph, nframes = N_frame, fps = 100)


# アニメーション用のサンプルの密度を作図
anime_prop_graph <- ggplot() + 
  geom_histogram(data = anime_freq_df, mapping = aes(x = lambda, y = ..density..), 
                 breaks = seq(from = min(lambda_vals), to = max(lambda_vals), length.out = 30), 
                 fill = "#00A968", color = "#00A968") + # ヒストグラム
  geom_line(data = anime_dens_df, mapping = aes(x = lambda, y = density), 
            color = "darkgreen", linetype = "dashed") + # 元の分布
  geom_point(data = anime_data_df, mapping = aes(x = lambda, y = 0), 
             color = "orange", size = 5) + # サンプル
  gganimate::transition_manual(parameter) + # フレーム
  ylim(c(-0.01, max(dens_df[["density"]]) * 2)) + # y軸の表示範囲
  labs(title = "Gamma Distribution", 
       subtitle = "{current_frame}") # ラベル

# gif画像を作成
gganimate::animate(anime_prop_graph, nframes = N_frame, fps = 100)


# 分布の生成 -------------------------------------------------------------------

### パラメータを生成 -----

# パラメータを指定
a <- 5
b <- 2

# サンプルサイズを指定
N <- 10

# ガウス分布・ポアソン分布のパラメータを生成
lambda_n <- rgamma(n = N, shape = a, rate = b)


### 分布の作図:(1次元ガウス分布) -----

# 平均パラメータを指定
mu = 0.0

# 精度パラメータの期待値を計算
E_lambda <- a / b

# 標準偏差の期待値を計算
E_sigma <- sqrt(1.0 / E_lambda)

# 作図用のxの点を作成
x_vals <- seq(from = mu - E_sigma*4, to = mu + E_sigma*4, length.out = 250)

# 精度パラメータの期待値による1次元ガウス分布を計算
res_dens_df <- tidyr::tibble(
  x = x_vals, # 確率変数
  density = dnorm(x = x_vals, mean = mu, sd = E_sigma), # 確率密度
  parameter = paste0("E[lambda]=", round(E_lambda, 2)) %>% 
    as.factor() # 色分け用のラベル
)

# lambdaの値ごとに分布を計算
for(lambda in lambda_n) {
  # 1次元ガウス分布を計算
  tmp_dens_df <- tidyr::tibble(
    x = x_vals, # 確率変数
    density = dnorm(x = x_vals, mean = mu, sd = sqrt(1 / lambda)), # 確率密度
    parameter = paste0("lambda=", round(lambda, 2)) %>% 
      as.factor() # 色分け用のラベル
  )
  
  # 結果を結合
  res_dens_df <- rbind(res_dens_df, tmp_dens_df)
}

# サンプルによる分布を作図
ggplot() + 
  geom_line(data = res_dens_df, 
            mapping = aes(x = x, y = density, color = parameter, alpha = parameter, linetype = parameter), size = 0.8) + # 分布
  scale_linetype_manual(values = c("dashed", rep("solid", times = N))) + # 線の種類
  scale_alpha_manual(values = c(1, rep(0.5, times = N))) + # 透過度
  labs(title = "Gaussian Distribution", 
       subtitle = paste0("a=", a, ", b=", b), 
       x = expression(lambda)) # ラベル


### 分布の作図:(ポアソン分布) -----

# パラメータの期待値を計算
E_lambda <- a / b

# 作図用のxの点を作成
x_vals <- seq(from = 0, to = ceiling(E_lambda) * 4)

# パラメータの期待値によるガンマ分布を計算
res_prob_df <- tidyr::tibble(
  x = x_vals, # 確率変数
  probability = dpois(x = x_vals, lambda = E_lambda), # 確率
  parameter = paste0("E[lambda]=", round(E_lambda, 2)) %>% 
    as.factor() # 色分け用のラベル
)

# lambdaの値ごとに分布を計算
for(lambda in lambda_n) {
  # ガンマ分布を計算
  tmp_prob_df <- tidyr::tibble(
    x = x_vals, # 確率変数
    probability = dpois(x = x_vals, lambda = lambda), # 確率
    parameter = paste0("lambda=", round(lambda, 2)) %>% 
      as.factor() # 色分け用のラベル
  )
  
  # 結果を結合
  res_prob_df <- rbind(res_prob_df, tmp_prob_df)
}

# サンプルによる分布を作図
ggplot() + 
  geom_step(data = res_prob_df, 
            mapping = aes(x = x, y = probability, color = parameter, alpha = parameter, linetype = parameter), 
            direction = "mid", size = 0.8) + # 分布
  #geom_bar(data = res_prob_df, mapping = aes(x = x, y = probability, color = parameter, linetype = parameter), 
  #         stat = "identity", alpha = 0) + # 分布
  scale_linetype_manual(values = c("dashed", rep("solid", times = N))) + # 線の種類
  scale_alpha_manual(values = c(1, rep(0.5, times = N))) + # 透過度
  scale_x_continuous(breaks = x_vals) + # x軸目盛
  labs(title = "Poisson Distribution", 
       subtitle = paste0("a=", a, ", b=", b), 
       x = expression(lambda)) # ラベル

File: code/Python/gamma.py
# ガンマ分布

# 利用するライブラリ
import numpy as np
from scipy.stats import gamma, norm, poisson # ガンマ分布, 1次元ガウス分布, ポアソン分布
import scipy.special as sp # ガンマ関数, 対数ガンマ関数
import matplotlib.pyplot as plt
from matplotlib.animation import FuncAnimation

#%%

### 確率密度の計算

# パラメータを指定
a = 2.0
b = 2.0

# 確率変数の値を指定
lmd = 2.0


# 定義式により確率密度を計算
C = b**a / sp.gamma(a)
dens = C * lmd**(a - 1.0) * np.exp(-b * lmd)
print(dens)

# 対数をとった定義式により確率密度を計算
log_C = a * np.log(b) - sp.loggamma(a)
log_dens = log_C + (a - 1.0) * np.log(lmd) - b * lmd
dens = np.exp(log_dens)
print(dens, log_dens)

# ガンマ分布の関数により確率密度を計算
dens = gamma.pdf(x=lmd, a=a, scale=1.0 / b)
print(dens)

# ガンマ分布の対数をとった関数により確率密度を計算
log_dens = gamma.logpdf(x=lmd, a=a, scale=1.0 / b)
dens = np.exp(log_dens)
print(dens, log_dens)

#%%

### 統計量の計算

# パラメータを指定
a = 2.0
b = 2.0


# 計算式により平均を計算
E_lambda = a / b
print(E_lambda)

# 計算式により分散を計算
V_lambda = a / b**2
print(V_lambda)

# 計算式により最頻値を計算
mode_lambda = (a - 1.0) / b
print(mode_lambda)

# ガンマ分布の関数により平均を計算
E_lambda = gamma.mean(a=a, scale=1.0 / b)
print(E_lambda)

# ガンマ分布の関数により分散を計算
V_lambda = gamma.var(a=a, scale=1.0 / b)
print(V_lambda)

#%%

### 分布の可視化

## 分布の計算

# パラメータを指定
a = 2.0
b = 2.0

# 作図用のlambdaの点を作成
lambda_vals = np.linspace(start=0.0, stop=5.0, num=250)

# ガンマ分布を計算
density = gamma.pdf(x=lambda_vals, a=a, scale = 1.0 / b)

#%%

## 分布の作図

# ガンマ分布を作図
plt.figure(figsize=(12, 9)) # 図の設定
plt.plot(lambda_vals, density, color='#00A968') # 折れ線グラフ
plt.xlabel('$\lambda$') # x軸ラベル
plt.ylabel('density') # y軸ラベル
plt.suptitle('Gamma Distribution', fontsize=20) # 全体のタイトル
plt.title('$a=' + str(a) + ', b=' + str(b) + '$', loc='left') # タイトル
plt.grid() # グリッド線
plt.show() # 描画

#%%

# 統計量を計算
E_lmd = a / b
s_lmd = np.sqrt(a / b**2)
mode_lmd = (a - 1.0) / b

# 統計量を重ねたガンマ分布を作図
plt.figure(figsize=(12, 9)) # 図の設定
plt.plot(lambda_vals, density, color='#00A968') # 分布
plt.vlines(x=E_lmd, ymin=0.0, ymax=np.max(density), color='orange', linestyle='--', label='$E[\lambda]$') # 平均
plt.vlines(x=E_lmd - s_lmd, ymin=0.0, ymax=np.max(density), color='orange', linestyle=':', label='$E[\lambda] - \\sqrt{V[\lambda]}$') # 平均 - 標準偏差
plt.vlines(x=E_lmd + s_lmd, ymin=0.0, ymax=np.max(density), color='orange', linestyle=':', label='$E[\lambda] + \\sqrt{V[\lambda]}$') # 平均 + 標準偏差
plt.vlines(x=mode_lmd, ymin=0.0, ymax=np.max(density), color='chocolate', linestyle='--', label='$mode[\lambda]$') # 最頻値
plt.xlabel('$\lambda$') # x軸ラベル
plt.ylabel('density') # y軸ラベル
plt.suptitle('Gamma Distribution', fontsize=20) # 全体のタイトル
plt.title('$a=' + str(a) + ', b=' + str(b) + '$', loc='left') # タイトル
plt.legend() # 凡例
plt.grid() # グリッド線
plt.show() # 描画

#%%

### パラメータと分布の形状の関係

## a, bの影響

# パラメータとして利用する値を指定
a_vals = np.arange(start=0.1, stop=10.1, step=0.1)
b_vals = np.arange(start=0.1, stop=10.1, step=0.1)
print(len(a_vals)) # フレーム数

# 固定するパラメータを指定
a = 2.0
b = 2.0

# 作図用のlambdaの点を作成
lambda_vals = np.linspace(start=0.0, stop=5.0, num=250)

# y軸(確率密度)の最大値を設定
dens_max = 4.0

# 図を初期化
fig = plt.figure(figsize=(12, 9)) # 図の設定
fig.suptitle('Gamma Distribution', fontsize=20) # 全体のタイトル

# 作図処理を関数として定義
def update(i):
    # 前フレームのグラフを初期化
    plt.cla()
    
    # i番目のパラメータを取得
    a = a_vals[i]
    #b = b_vals[i]
    
    # ガンマ分布を計算
    density = gamma.pdf(x=lambda_vals, a=a, scale = 1.0 / b)
    
    # ガンマ分布を作図
    plt.plot(lambda_vals, density, color='#00A968') # 折れ線グラフ
    plt.xlabel('$\lambda$') # x軸ラベル
    plt.ylabel('density') # y軸ラベル
    plt.title('$a=' + str(np.round(a, 1)) + ', b=' + str(np.round(b, 1)) + '$', loc='left') # タイトル
    plt.grid() # グリッド線
    plt.ylim(ymin=-0.1, ymax=dens_max) # y軸の表示範囲

# gif画像を作成
anime_dens = FuncAnimation(fig, update, frames=len(a_vals), interval=100)

# gif画像を保存
anime_dens.save('ProbabilityDistribution/Gamma_dens.gif')

#%%

### 乱数の生成

## 乱数の生成

# パラメータを指定
a = 2.0
b = 2.0

# データ数(サンプルサイズ)を指定
N = 1000

# ガンマ分布に従う乱数を生成
lambda_n = np.random.gamma(shape=a, scale=1.0 / b, size=N)

# 作図用のlambdaの点を作成
lambda_vals = np.linspace(start=0.0, stop=np.max(lambda_n) + 1.0, num=250)

# ガンマ分布を計算
density = gamma.pdf(x=lambda_vals, a=a, scale=1.0 / b)

#%%

## 乱数の可視化

# サンプルのヒストグラム(頻度)を作成
plt.figure(figsize=(12, 9)) # 図の設定
plt.hist(x=lambda_n, bins=30, range=(lambda_vals.min(), lambda_vals.max()), color='#00A968') # ヒストグラム
plt.xlabel('$\lambda$') # x軸ラベル
plt.ylabel('density') # y軸ラベル
plt.suptitle('Gamma Distribution', fontsize=20) # 全体のタイトル
plt.title('$a=' + str(a) + ', b=' + str(b) + ', N=' + str(N) + '$', loc='left') # タイトル
plt.grid() # グリッド線
plt.ylim(ymin=-0.01) # y軸の表示範囲
plt.show() # 描画

# サンプルのヒストグラム(密度)を作成
plt.figure(figsize=(12, 9)) # 図の設定
plt.hist(x=lambda_n, bins=30, range=(lambda_vals.min(), lambda_vals.max()), density=True, color='#00A968') # ヒストグラム
plt.plot(lambda_vals, density, color='green', linestyle='--') # 元の分布
plt.xlabel('$\lambda$') # x軸ラベル
plt.ylabel('density') # y軸ラベル
plt.suptitle('Gamma Distribution', fontsize=20) # 全体のタイトル
plt.title('$a=' + str(a) + ', b=' + str(b) + ', N=' + str(N) + '$', loc='left') # タイトル
plt.grid() # グリッド線
plt.ylim(ymin=-0.01) # y軸の表示範囲
plt.show() # 描画

#%%

## アニメーションによる可視化:(頻度)

# フレーム数を指定
N_frame = 100

# 図を初期化
fig = plt.figure(figsize=(12, 9)) # 図の設定
fig.suptitle('Gaussian Distribution', fontsize=20) # 全体のタイトル

# y軸(頻度)の最大値を設定
freq_max = np.max(
    np.histogram(a=lambda_n[:N_frame], bins=30, range=(lambda_vals.min(), lambda_vals.max()))[0], 
) + 1.0

# 作図処理を関数として定義
def update(n):
    # 前フレームのグラフを初期化
    plt.cla()
    
    # サンプルのヒストグラムを作成
    plt.hist(x=lambda_n[:(n+1)], bins=30, range=(lambda_vals.min(), lambda_vals.max()), color='#00A968', zorder=1) # ヒストグラム
    plt.scatter(x=lambda_n[n], y=0.0, s=100, color='orange', zorder=2) # サンプル
    plt.xlabel('$\lambda$') # x軸ラベル
    plt.ylabel('frequency') # y軸ラベル
    plt.suptitle('Gamma Distribution', fontsize=20) # 全体のタイトル
    plt.title('$a=' + str(a) + ', b=' + str(b) + ', N=' + str(n + 1) + '$', loc='left') # タイトル
    plt.grid() # グリッド線
    plt.ylim(ymin=-0.5, ymax=freq_max) # y軸の表示範囲

# gif画像を作成
anime_freq = FuncAnimation(fig, update, frames=N_frame, interval=100)

# gif画像を保存
anime_freq.save('ProbabilityDistribution/Gamma_freq.gif')

#%%

## アニメーションによる可視化:(密度)

# フレーム数を指定
N_frame = 100

# 図を初期化
fig = plt.figure(figsize=(12, 9)) # 図の設定
fig.suptitle('Gaussian Distribution', fontsize=20) # 全体のタイトル

# y軸(確率密度)の最大値を設定
dens_max = np.max(
    np.hstack([
        np.histogram(a=lambda_n[:N_frame], bins=30, range=(lambda_vals.min(), lambda_vals.max()), density=True)[0], 
        density
    ])
) + 0.1

# 作図処理を関数として定義
def update(n):
    # 前フレームのグラフを初期化
    plt.cla()
    
    # サンプルのヒストグラムを作成
    plt.hist(x=lambda_n[:(n+1)], bins=30, range=(lambda_vals.min(), lambda_vals.max()), density=True, color='#00A968', zorder=1) # ヒストグラム
    plt.plot(lambda_vals, density, color='green', linestyle='--', zorder=2) # 元の分布
    plt.scatter(x=lambda_n[n], y=0.0, s=100, color='orange', zorder=3) # サンプル
    plt.xlabel('$\lambda$') # x軸ラベル
    plt.ylabel('density') # y軸ラベル
    plt.suptitle('Gamma Distribution', fontsize=20) # 全体のタイトル
    plt.title('$a=' + str(a) + ', b=' + str(b) + ', N=' + str(n + 1) + '$', loc='left') # タイトル
    plt.grid() # グリッド線
    plt.ylim(ymin=-0.01, ymax=dens_max) # y軸の表示範囲

# gif画像を作成
anime_freq = FuncAnimation(fig, update, frames=N_frame, interval=100)

# gif画像を保存
anime_freq.save('ProbabilityDistribution/Gamma_prop.gif')

#%%

### 分布の生成

## パラメータの生成

# パラメータを指定
a = 5.0
b = 2.0

# サンプルサイズを指定
N = 10

# ガウス分布・ポアソン分布のパラメータを生成
lambda_n = np.random.gamma(shape=a, scale=1.0 / b, size=N)

#%%

## 分布の作図:(1次元ガウス分布)

# 平均パラメータを指定
mu = 0.0

# 精度パラメータの期待値を計算
E_lambda = a / b

# 標準偏差の期待値を計算
E_sigma = np.sqrt(1.0 / E_lambda)

# 作図用のxの点を作成
x_vals = np.linspace(start=mu - E_sigma*4.0, stop=mu + E_sigma*4.0, num=250)

# 精度パラメータの期待値による1次元ガウス分布を計算
E_dens = norm.pdf(x=x_vals, loc=mu, scale=E_sigma)

# サンプルによる分布を作図
plt.figure(figsize=(12, 9)) # 図の設定
plt.plot(x_vals, E_dens, color='blue', linestyle='--', label='$E[\lambda]=' + str(np.round(E_lambda, 2)) + '$') # 期待値による分布
for n in range(N):
    tmp_dens = norm.pdf(x=x_vals, loc=mu, scale=np.sqrt(1.0 / lambda_n[n]))
    plt.plot(x_vals, tmp_dens, alpha=0.5, label='$\lambda=' + str(np.round(lambda_n[n], 2)) + '$') # サンプルによる分布
plt.xlabel('x') # x軸ラベル
plt.ylabel('density') # y軸ラベル
plt.suptitle('Gaussian Distribution', fontsize=20) # 全体のタイトル
plt.title('$a=' + str(a) + ', b=' + str(b) + '$', loc='left') # タイトル
plt.legend() # 凡例
plt.grid() # グリッド線
plt.show() # 描画

#%%

## 分布の作図:(ポアソン分布)

# パラメータの期待値を計算
E_lambda = a / b

# 作図用のxの点を作成
x_vals = np.arange(np.ceil(E_lambda) * 4.0)

# パラメータの期待値によるポアソン分布を計算
E_prob = poisson.pmf(k=x_vals, mu=E_lambda)

# サンプルによる分布を作図
plt.figure(figsize=(12, 9)) # 図の設定
plt.step(x=x_vals, y=E_prob, where='mid', 
         color='blue', linestyle='--', label='$E[\lambda]=' + str(np.round(E_lambda, 2)) + '$') # 期待値による分布
for n in range(N):
    tmp_prob = poisson.pmf(k=x_vals, mu=lambda_n[n])
    plt.step(x=x_vals, y=tmp_prob, where='mid', 
             alpha=0.5, label='$\lambda=' + str(np.round(lambda_n[n], 2)) + '$') # サンプルによる分布
plt.xlabel('x') # x軸ラベル
plt.ylabel('density') # y軸ラベル
plt.suptitle('Poisson Distribution', fontsize=20) # 全体のタイトル
plt.title('$a=' + str(a) + ', b=' + str(b) + '$', loc='left') # タイトル
plt.legend() # 凡例
plt.grid() # グリッド線
plt.show() # 描画

#%%

